compiler/rustc_hir_analysis/src/coherence/inherent_impls.rs

```diff
@@ -112,6 +112,41 @@ impl<'tcx> InherentCollect<'tcx> {
         }
     }
 
+    fn check_field_type(&mut self, container: Ty<'tcx>, field_path: &'tcx List<FieldPathSegment>) {
+        let mut cur = container;
+        let mut idxs = Vec::new();
+
+        let infcx_;
+        let infcx = if let Some(infcx) = self.infcx() {
+            infcx
+        } else {
+            assert!(!qself_ty.has_infer());
+            infcx_ = tcx.infer_ctxt().build(TypingMode::non_body_analysis());
+            &infcx_
+        };
+
+        'outer: for field in fields {
+            cur = infcx_.shallow_resolve(cur);
+            match cur.kind() {
+                ty::Adt(def, args) if def.is_struct() => {
```

Comment:
I bumped into the inherent and incoherent `impl` question again, which has connection to method probing.

So I definitely agree with your rule that a `impl`. I don't think that it is safe for now to allow users to write `impl [Trait for] field_of!(..)` easily when types, other than one from the last segment, are defined in remote crates. I doubt such `impl`s deliver as much value as `impl` on `Box<_>` at the moment unless real necessity arises. So your plan sounds all good to me.